core/src/hir/types.rs

```diff
@@ -23,6 +23,14 @@ pub enum Type<P: TyPosition = Everywhere> {
     Enum(EnumPath),
     Slice(Slice),
     Callback(P::CallbackInstantiation), // only a Callback if P == InputOnly
+    /// `DiplomatOption<T>`, for  a primitive, struct, or enum `T`.
+    ///
+    /// In some cases this can be specified as `Option<T>`, but under the hood it gets translated to
+    /// the DiplomatOption FFI-compatible union type.
+    ///
+    /// This is *different* from `Option<&T>` and `Option<Box<T>` for an opaque `T`: That is represented as
+    /// a nullable pointer.
+    DiplomatOption(Box<Type<P>>),
```

Comment:
Between this and `ReturnType::Nullable` we now have two ways to model optional returns...

Reply:
Correct, however this should never occur in a return. I'll add a comment.